.github/copilot-instructions.md

```diff
@@ -0,0 +1,314 @@
+1. Browse to http://localhost:3020
+2. Test user registration/login flow
+3. Navigate key sections (tracks, exercises, community)
+4. Test JavaScript interactions (code editor, modals)
+5. Verify CSS styling renders correctly
+
+### Pre-commit Validation
+ALWAYS run before committing:
+```bash
+bundle exec rubocop --except Metrics
+yarn test
+bundle exec rails test:zeitwerk
+```
+
+## Critical Warnings and Limitations
+
+### Build Timing - NEVER CANCEL
+- **Bundle install: 15-20 minutes** - Native extensions for grpc, skylight, commonmarker take time
+- **Ruby test suite: 10-15 minutes** - Full suite with database operations
+- **System tests: 15-20 minutes** - Browser automation with Capybara/Selenium
+- **Asset builds: 2-5 minutes each** - CSS and JavaScript compilation
+
+### Authentication Requirements
+- **NPM Token required**: Private package access needs `NPM_TOKEN` environment variable
+- **AWS LocalStack**: Required for S3, DynamoDB, and other AWS service mocking
+- **Database credentials**: Must match config/database.yml settings
+
+### Version Dependencies (CRITICAL)
+- **Ruby 3.4.4**: Exact version required - Gemfile.lock specifies this
+- **Node.js 20+**: Required for esbuild and modern JS features
+- **MySQL 5.7**: Database schema depends on this version
+- **Bundler 2.6.9**: Locked version in Gemfile.lock
+
+### Known Limitations
+- Ruby version managers (rbenv/rvm) recommended for version management
+- Some JS packages require private NPM access (will fail in some environments)  
+- Docker services must be running before Rails server starts
+- Database must be properly configured with utf8mb4 collation
+- Large codebase - initial setup can take 30+ minutes total
+
+## Important Directories
+
+### Application Structure
+- `app/controllers/` - Rails controllers
+- `app/models/` - ActiveRecord models and business logic
+- `app/commands/` - Mandate command objects for business logic (see Command Pattern below)
+- `app/javascript/` - React/TypeScript frontend code
+- `app/css/` - PostCSS stylesheets with Tailwind
+- `app/views/` - HAML view templates
+- `test/` - All test files (unit, integration, system)
+- `test/system/` - Capybara system tests
+- `test/javascript/` - Jest test files
+
+### Command Pattern with Mandate
+The `/app/commands` directory contains command objects that encapsulate business logic using the Mandate gem. This pattern promotes:
+- Single Responsibility Principle
+- Testable business logic separation from controllers
+- Consistent error handling and callbacks
+
+#### Command Structure
+Commands follow a consistent pattern:
+
+```ruby
+class User::Update
+  include Mandate
+  include Mandate::Callbacks  # Optional, for success/failure callbacks
+
+  initialize_with :user, :params  # Defines constructor parameters
+
+  def call  # Single method that performs the operation
+    # Business logic here
+    abort!(errors) if errors.present?  # Stops execution and triggers failure callback
```

Comment:
`abort!` isn't a special thing. It would just have been method in this command, like `guard!` or similar. In this situation you probably want to raise an exception.